### PR TITLE
369 name weave peers

### PR DESCRIPTION
**Pull request by awh**
Implement `-hostname` flag to weave router for inclusion in `weave status`:

    root@weave01:~# weave status
    weave router git-170724472555
    Encryption off
    Our name is 7a:85:0d:3f:4b:8b (weave01)
    Sniffing traffic on &{25 65535 ethwe 26:47:3f:b1:9f:35 up|broadcast|multicast}
    MACs:
    e6:eb:3d:c2:f2:c0 -> 7a:85:0d:3f:4b:8b (2015-02-19 14:36:43.698599704 +0000 UTC)
    ea:9b:f6:9a:a8:94 -> 7a:ee:54:05:6d:19 (2015-02-19 14:36:45.467878022 +0000 UTC)
    02:c5:dc:fa:4f:19 -> 7a:c4:7d:e3:0f:9b (2015-02-19 14:36:50.594692416 +0000 UTC)
    Peers:
    Peer 7a:c4:7d:e3:0f:9b (weave03) (v37) (UID 12377675045837118970)
       -> 7a:ee:54:05:6d:19 (weave02) [178.62.77.81:6783]
       -> 7a:85:0d:3f:4b:8b (weave01) [178.62.57.35:6783]
    Peer 7a:85:0d:3f:4b:8b (weave01) (v8) (UID 8327482320931219319)
       -> 7a:c4:7d:e3:0f:9b (weave03) [178.62.106.102:59473]
       -> 7a:ee:54:05:6d:19 (weave02) [178.62.77.81:48078]
    Peer 7a:ee:54:05:6d:19 (weave02) (v4) (UID 13020696009711836879)
       -> 7a:85:0d:3f:4b:8b (weave01) [178.62.57.35:6783]
       -> 7a:c4:7d:e3:0f:9b (weave03) [178.62.106.102:53570]

Notes:

* `weave launch` does not accept user supplied `-hostname` as this would require introducing `getopt` or similar into the bash script and given [#307](https://github.com/zettio/weave/issues/307) is in the offing it didn't seem prudent; script supplies kernel hostname directly to weave router for now. Please advise if you would like this addressed before merging
* Protocol version was increased due to binary incompatibility introduced by inclusion of hostname in the gossip protocol messages

### CODE REVIEW

**Comment by rade** [comment]
Just for reference, this is meant to address #369.

**Comment by rade** [comment]
Referring to this new piece of data as 'hostname' is wrong; it could be anything, and in many cases will *not* be the hostname. Yes, I know #369 calls it 'hostname', but I didn't file that issue ;)

And yes, we really, really must be able to pass that thing into `weave launch`.

**Comment by awh** [comment]
Understood. Do you have a preference on what you'd like it to be called?

**Comment by rade** [comment]
> Do you have a preference on what you'd like it to be called?

I am sure I do, but I don't know what it is :)

**Comment by rade** [comment]
'displayname'? Not great, so happy to consider alternatives.

File: weaver/main.go [hunk 1]
```diff
@@ -87,6 +89,18 @@ func main() {
 		routerName = iface.HardwareAddr.String()
 	}
 
+	if hostName == "" {
+		osHostName, err := os.Hostname()
+		if err != nil {
+			log.Fatal(err)
+		}
+		if osHostName == "" {
+			hostName = "<unknown>"
```
Comment (rade) [inline weaver/main.go]: I reckon that will look rather odd in the various places where that gets displayed. Set it to "", perhaps?

**Comment by awh** [comment]
I was just about to suggest that or friendlyname, informalname....

File: weaver/main.go [hunk 2]
```diff
@@ -121,8 +135,8 @@ func main() {
 		defer profile.Start(&p).Stop()
 	}
 
-	router := weave.NewRouter(iface, ourName, []byte(password), connLimit, bufSz*1024*1024, logFrame)
-	log.Println("Our name is", router.Ourself.Name)
+	router := weave.NewRouter(iface, ourName, hostName, []byte(password), connLimit, bufSz*1024*1024, logFrame)
+	log.Println("Our name is", router.Ourself.Name, "(" + hostName + ")")
```
Comment (rade) [inline weaver/main.go]: IMO it would be "more correct" and consistent to get extract the info from router.Ourself.

**Comment by rade** [comment]
'friendlyname' is quite nice. let's invite @bboreham to help paint this particular bike shed :)

File: weaver/main.go (hunk 2 above)
Comment (awh) [inline weaver/main.go]: Actually this logging is duplicated in router.Status() - how about log.Println(router.Status()) instead?

**Comment by dpw** [comment]
If I may join the bike-shedding, what about "nickname"?

File: weaver/main.go (hunk 2 above)
Comment (rade) [inline weaver/main.go]: That's by coincidence. And it also prints a lot more, which isn't particularly meaningful at this stage.

**Comment by rade** [comment]
I do like 'nickname'.

File: weaver/main.go (hunk 2 above)
Comment (awh) [inline weaver/main.go]: Hmm that won't look great - router.Status() returns a single string with embedded newlines...

**Comment by awh** [comment]
Addressed comments to date!

**Comment by rade** [comment]
The new option needs to added to the docs. Somewhere.

**Comment by awh** [comment]
Just thinking about the default value of the nickname - would `$(hostname -f)` be better than `$(hostname)`?

**Comment by rade** [comment]
Ouch, `hostname` isn't even in POSIX! I guess we don't care, but still surprising.

**Comment by awh** [comment]
That is surprising. Bash has a `HOSTNAME` environment variable, however the script depends on `/bin/sh`. Unfortunately we can't solely depend on `os.Hostname()` from Go because it obtains the hostname of the container and not of the underlying host kernel :(

With respect to the other issue (short hostname vs FQDN) I suspect the latter is going to be useful if folks have naming schemes like `app01.dc01.example.com`, `app01.dc02.example.com`...

**Comment by rade** [comment]
Can anybody think of a *downside* to using FQDNs? @bboreham

**Comment by bboreham** [comment]
When I thought of this originally, I meant the "hostname" of the container.  So we didn't have to implement any flag-parsing, because `--hostname` is already a Docker option.  I'm not clear why you went down this other route.

**Comment by rade** [comment]
The hostname of the weave container is not exactly meaningful.

**Comment by bboreham** [comment]
It defaults to the container ID, which I usually know when testing, and I can set to anything else I feel like.  So it is meaningful to me.

**Comment by rade** [comment]
setting it to anything else is not exactly straightforward

**Comment by bboreham** [comment]
I seem to be missing this part.  Extra args in `weave launch` are handed to Docker; I can add `--hostname` to those args.  Where does it get difficult?

**Comment by rade** [comment]
extra args to `weave launch` are handed to weave.

**Comment by bboreham** [comment]
Aha.  I think that belongs as a comment to #369.

**Comment by rade** [comment]
you'd have to use `WEAVE_DOCKER_ARGS`.

**Comment by awh** [comment]
The only downside I can think of to FQDNs is that it might cause the output of `weave status` to wrap if you have a long domain...

**Comment by bboreham** [comment]
I just took a look at `hostname -f` on one of the machines I use for testing, and it is "bb-ub14-test2.bb-ub14-test2.f3.internal.cloudapp.net".  I think I would hate that more than the hex numbers I have to work with today.

**Comment by rade** [comment]
and the alternative is what?

**Comment by bboreham** [comment]
My idea was for the user (me) to specify the name with a command-line flag.

**Comment by rade** [comment]
Which you can do. The debate re fqdn vs unqualified concerns what should happen when you *don't* specify a name.

**Comment by bboreham** [comment]
From my sampling of two Cloud VMs, `hostname` seems to be less worse than `hostname -f`.  The other example is "ip-172-31-20-70.eu-west-1.compute.internal"

Given the option to supply on the command line, anyone that cares to can use `hostname -f` to supply that, right?

**Comment by rade** [comment]
I don't understand why you feel it's better to default to the short names instead of the the FQDNs. The latter at least give you a clue what cloud the host is in.

**Comment by bboreham** [comment]
Based on those two samples, I can tell what cloud the host is in from the short name.

**Comment by rade** [comment]
> Based on those two samples, I can tell what cloud the host is in from the short name.

Perhaps, but in case of EC2, the FQDN also tells you the region, which is useful.

What's your objection to using the FQDN by default? That it can be quite long and thus clutter the output? I can see how that could be annoying, but IMO the default should favour uniqueness & more info over compactness. As you say, it's easy to override by supplying `-hostname $(hostname)`.

File: site/troubleshooting.md [hunk 3]
```diff
@@ -86,7 +86,11 @@ The terms used here are explained further at
 [how it works](how-it-works.html).
 
 The 'Our name' line identifies the local weave router as a peer in the
-weave network.
+weave network. It displays both the peer name and the peer's nickname
+in parenthesis; the peer name defaults to a MAC address, whilst the
```
Comment (rade) [inline site/troubleshooting.md]: We don't provide an option to override the peer name, so saying that it "defaults to" a MAC is wrong. Also "a MAC"? *what* MAC? Yes, how-it-works (which, incidentally, needs updating to mention the nickname) contains similar wording but the context there is the name *format*, i.e. "defaults to a MAC" means "the default format of a name is MAC" (I'm not exactly delighted with that phrasing either though).

File: weaver/main.go [hunk 4]
```diff
@@ -87,6 +89,14 @@ func main() {
...
 	}
 
+	if nickName == "" {
+		osHostname, err := os.Hostname()
+		if err != nil {
+			log.Fatal(err)
+		}
+    	nickName = osHostname
+	}
+
```
Comment (rade) [inline weaver/main.go]: The indentation has gone rather wrong in the above. Also, might as well assign straight to nickName.

File: weaver/main.go [hunk 5]
```diff
@@ -49,6 +50,7 @@ func main() {
 	flag.BoolVar(&justVersion, "version", false, "print version and exit")
 	flag.StringVar(&ifaceName, "iface", "", "name of interface to read from")
 	flag.StringVar(&routerName, "name", "", "name of router (defaults to MAC)")
+	flag.StringVar(&nickName, "nickname", "", "nickname of host (defaults to container hostname)")
```
Comment (rade) [inline weaver/main.go]: You can't mention 'container' here; this code has no idea it's running as a container, and indeed might not.

**Comment by rade** [comment]
IF we are going to go with `hostname -f` as the default, then the go-level default should also be the FQDN. How hard is that?

**Comment by awh** [comment]
Go question - which is preferred?
As the code is now:
````
	if nickName == "" {
		osHostname, err := os.Hostname()
		if err != nil {
			log.Fatal(err)
		}
		nickName = osHostname
	}
````
Assigning directly to nickName (NB using '=' instead of ':='); now we're depending upon and overwriting an earlier 'err' declaration:
````
	if nickName == "" {
		nickName, err = os.Hostname()
		if err != nil {
			log.Fatal(err)
		}
	}
````
Explicitly shadow err:
````
	if nickName == "" {
		err := nil
		nickName, err = os.Hostname()
		if err != nil {
			log.Fatal(err)
		}
	}
````
Unfortunately this doesn't work because the inner `nickName` shadows the outer:
````
	if nickName == "" {
		nickName, err := os.Hostname()
		if err != nil {
			log.Fatal(err)
		}
	}
````

**Comment by awh** [comment]
Getting the FQDN in pure Go is doable:

````
func main() {
	hostname, err := os.Hostname()
	if err == nil {
		addrs, err := net.LookupHost(hostname)

		if err == nil {
			names, err := net.LookupAddr(addrs[0])

			if err == nil {
				fmt.Printf("%v", names)
			}
		}
	}
}
````
(Obviously both `LookupHost` and `LookupAddr` can return zero to many items, which the above code doesn't handle). In my Vagrant VM the above code prints this:
````
vagrant@ubuntu-14:~$ ./main
[ubuntu-14.04-amd64-vbox ubuntu-14]
````
Which looks to me like the `LookupAddr` function isn't parsing out the alias from the `hosts` file:
````
vagrant@ubuntu-14:~$ cat /etc/hosts
127.0.0.1	localhost
127.0.1.1	ubuntu-14.04-amd64-vbox	ubuntu-14
````

**Comment by rade** [comment]
> Go question - which is preferred?

The 2nd one. Nothing wrong with re-using `err`.

> Getting the FQDN in pure Go is doable.

That's horrific. going via reverse lookup. Yuck.

**Comment by awh** [comment]
:) As far as I am aware, this is exactly what `hostname -f` is doing underneath, albeit via `libnsl` to parse `nsswitch.conf`, `resolv.conf` and `hosts`...

**Comment by awh** [comment]
Shelling out to `hostname -f` from Go is an option I suppose :)

**Comment by rade** [comment]
> Shelling out to `hostname -f` from Go is an option I suppose :)

Sure. If we want to undo all the work we've done to avoid using the native resolver libraries. Oh and to producer an image containing just a single executable.

I'm afraid `hostname`/`os.Hostname()` is doing to have to be the default.

**Comment by awh** [comment]
> Sure. If we want to undo all the work we've done to avoid using the native resolver libraries. Oh and to producer an image containing just a single executable.

I was half joking, but learned something useful! Is the desire to avoid the resolver related to the warnings in #5?

> I'm afraid hostname/os.Hostname() is doing to have to be the default.

In that case as far as I am aware the currently implemented behaviour matches the now finalised requirements...

**Comment by rade** [comment]
>  Is the desire to avoid the resolver related to the warnings in #5?

Yes. But off-topic.

File: router/gossip_test.go [hunk 6]
```diff
@@ -19,7 +19,7 @@ type mockChannelConnection struct {
 // We need to create some dummy channels otherwise tests hang on nil
 // channels when Router.OnGossip() calls async methods.
 func NewTestRouter(name PeerName) *Router {
-	router := NewRouter(nil, name, nil, 10, 1024, nil)
+	router := NewRouter(nil, name, "hostname", nil, 10, 1024, nil)
```
Comment (rade) [inline router/gossip_test.go]: I reckon "" makes a better test nickname here and elsewhere.

File: weave [hunk 7]
```diff
@@ -475,18 +475,31 @@ case "$COMMAND" in
...
+                    ;;
+            esac
+        done
         # Set WEAVE_DOCKER_ARGS in the environment in order to supply
         # additional parameters, such as resource limits, to docker
         # when launching the weave container.
         CONTAINER=$(docker run --privileged -d --name=$CONTAINER_NAME \
             -p $PORT:$PORT/tcp -p $PORT:$PORT/udp -e WEAVE_PASSWORD \
-            $WEAVE_DOCKER_ARGS $IMAGE -name $MACADDR -iface $CONTAINER_IFNAME "$@")
+            $WEAVE_DOCKER_ARGS $IMAGE -name $MACADDR -iface $CONTAINER_IFNAME -nickname "$NICKNAME" "$@")
```
Comment (rade) [inline weave]: Nitpick: The arg order here should be: iface,name,nickname. To match the order of the arg parsing declaration and process in main.go, and the order of args in NewRouter.

**Comment by awh** [comment]
Above commits address comments to date.

**Comment by rade** [comment]
You missed a fair few places where "hostname" is used as a nickname in tests.

**Comment by awh** [comment]
Apologies, I neglected to look further a field than the specific file you commented on. Fixed.
